[clang][ARM] disable frame pointers by default for bare metal ARM targets

stuij opened the pull request:
because:
- This brings Clang in line with GCC for which this is the default for ARM
- It frees up a register, so performance increase, especially on Thumb/6-M
- It will decrease code size

stuij committed:
[clang][ARM] disable frame pointers by default for bare metal ARM targets

because:
- This brings Clang in line with GCC for which this is the default for ARM
- It frees up a register, so performance increase, especially on Thumb/6-M
- It will also decrease code size

stuij commented:
This change was discussed in the LLVM Embedded Toolchains working group sync up two weeks ago, and at the time people were in agreement that it made sense:

https://discourse.llvm.org/t/llvm-embedded-toolchains-working-group-sync-up/63270/73#p-334070-peter-frame-pointers-on-or-off-by-default-7

stuij committed:
fix clang -> toolchains closing brace comment mislabel

stuij commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -148,6 +151,9 @@ static bool isARMBareMetal(const llvm::Triple &Triple) {
 
   return true;
 }
+} // namespace clang
```
thanks! done!

nickdesaulniers commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -151,6 +152,10 @@ static bool useFramePointerForTargetByDefault(const llvm::opt::ArgList &Args,
     }
   }
 
+  if (toolchains::isARMBareMetal(Triple)) {
+    return false;
+  }
```
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

stuij replied:
done!

stuij committed:
Remove braces around single statement `if` body

jroelofs commented on clang/lib/Driver/ToolChains/CommonArgs.cpp:
```diff
@@ -151,6 +152,9 @@ static bool useFramePointerForTargetByDefault(const llvm::opt::ArgList &Args,
     }
   }
 
+  if (toolchains::isARMBareMetal(Triple))
```
mind adding `&& Triple.getVendor() != Triple::VendorType::Apple` + some tests, so this doesn't change the default for Apple firmwares?

cc @kubamracek

stuij replied:
Hi @jroelofs, adding the check for Apple shouldn't be necessary. `isARMBareMetal` is trying to stay in it's lane by only returning true if the vendor is `none` and the environment is `eabi` or `eabihf`.

From grepping the LLVM tests I understand the Apple firmware triple would be something like `arm/thumb-apple-none-macho`, correct? I've added a number of tests to cover this.

stuij replied:
Perhaps renaming `isARMBareMetal` to `isEABIBareMetal` would make things clearer?

stuij replied:
`isArmEABIBareMetal`?

jroelofs replied:
good idea re: `isArmEABIBareMetal`

stuij replied:
done!

nickdesaulniers reviewed (commented):
What are your thoughts on moving the definition into the BareMetal.h header, then adding cases for arm, armeb, thumb, and thumbeb to the existing getArch switch in CommonArgs.cpp#L85? Then we should be able to inline isARMBareMetal and elide the getArch checks in it.

stuij commented:
I feel that in `useFramePointerForTargetByDef` fn, in general things are predicated first on platform and then on arch. That L85 arch switch statement seems to serve as a "these arches don't do anything special" early return.

It's up to interpretation, but putting the different Arm variants there seems less elegant to me (that might not be saying much..). I'd say let's leave it alone for now, and if the bare metal logic gets more impenetrable at some point, refactor.

stuij committed:
Add Apple embedded tests

also add a few more non-apple embedded tests and make isARMBareMetal slightly
more elegant

stuij committed:
satisfy clang-format

stuij committed:
Rename isARMBareMetal -> isArmEABIBareMetal

To make it clearer what the fn is testing for.

petrhosek commented on clang/lib/Driver/ToolChains/BareMetal.cpp:
```diff
@@ -128,12 +128,14 @@ BareMetal::BareMetal(const Driver &D, const llvm::Triple &Triple,
-      Triple.getArch() != llvm::Triple::thumb &&
-      Triple.getArch() != llvm::Triple::armeb &&
-      Triple.getArch() != llvm::Triple::thumbeb)
+bool isARMEABIBareMetal(const llvm::Triple &Triple) {
```
I'd consider moving this function to https://github.com/llvm/llvm-project/blob/c16c0e26236d0468a2411cb3bce219d3c70650e9/clang/lib/Driver/ToolChains/Arch/ARM.cpp, I think that would be a more appropriate location since that's where we keep other similar utility functions.

stuij replied:
Yes, that makes total sense. Done!

stuij committed:
Move isARMEABIBareMetal to clang/lib/Driver/ToolChains/Arch/ARM.cpp

asb commented:
I think this change in behaviour deserves an entry in clang/docs/ReleaseNotes.rst
